### ISSUE TITLE
Group/filter findings in different categories

### ISSUE THREAD

**Issue by n614cd** [semgrep/semgrep#8074]
**Is your feature request related to a problem? Please describe.**
No.

**Describe the solution you'd like**
Checkstyle and many similar tools do a very good job of helping maintain consistent code readability and the resulting quality. Increasing the code quality with more readable code reduces human errors and the chances for security mistakes. It would be great for Semgrep to take a more holistic view and include this aspect, including the ability to use these tools to help encourage/enforce company coding standards. 
This can be done through two approaches. One, a whole new category of issues, or expand the existing issues with a new priority of "code quality" instead of high/medium/low.

**Describe alternatives you've considered**
Other products, or hack my own solution.

**Use case**
Improve code quality, provide feedback to developers as part of the MR process when company standards have not been followed.

**Comment by r2c-demo** [comment]
This issue is synced in Linear at https://linear.app/semgrep/issue/PA-2867/code-qualityreadability. _Note_: this link is for r2c use only and is not accessible publicly.

**Comment by aryx** [comment]
I'm not sure what you want exactly, but if your "code quality improvment" can be written as a semgrep rule, then we will support it! If it's about fixing the indentation of your code, we rely on external tools like gofmt instead.

**Comment by n614cd** [comment]
Very likely could be converted to semgrep rules (I have not yet had time to read your rule documentation); but these are not security issues. Hence why I would hesitate to have them lumped into the existing High, Medium, Low of security findings. There is some formatting rules; however that is minor, and I tend to leave that to the IDEs to fix. Here is a parser and the rule set which is very popular in the Java world:
https://checkstyle.org/checks.html
Checkstyle goes into the following general critiques of code (my opinion):
1. Design problems from such things as complexity, common logic errors, missing conditions. A few examples: CyclomaticComplexity, ClassFanOutComplexity, AvoidInlineConditionals, FallThrough
2. Code Conventions (where things should be declared, how they should be declared...). A few examples: DeclarationOrder, DefaultComesLast, AvoidStarImport
3. Formatting issues.

Tim

**Comment by aryx** [comment]
Semgrep is not intended only for security issues. Its target user are also regular developers,
and you don't have to provide the high/medium/low metadata; metadata are optional.
Only severity is mandatory.

**Comment by aryx** [comment]
Stuff that can be encoded by a semgrep rule, we will support; other, probably not (unless we extend the expressive power of semgrep rules to be able to address such issues).

**Comment by n614cd** [comment]
@aryx 

It could be me, and how I have dealt with developers and these types of issues. 
But I have always differentiated between security issues, code convention issues, code coverage and testing issues.
Based on the UI for semgrep, I have not seen such differentiation between types of findings. This is critical from a project management perspective, since security and testing findings are often blockers; but other types are not.

**Comment by aryx** [comment]
In Semgrep Code there is a bit such a differentiation with the blocking vs MR vs audit rules in the rule board.

**Comment by n614cd** [comment]
How do you differentiate on the findings screen?

**Comment by aryx** [comment]
cc @raghavjain3

**Comment by nmote** [comment]
I think the confusion here is because this is feedback about how rules are organize on Semgrep App, and not about Semgrep itself.

**Comment by n614cd** [comment]
@nmote 

Not at all. Semgrep has a single set of rules, and a single set of findings. 
Checkstyle and other code quality tools use a very similar process as security scanners, hence why I think the Checkstyle rules could likely be ported to Semgrep. However, the issues found by Checkstyle are not security findings in the traditional sense.
For comparison, look at SonarQube. They have separate issues for Security, versus code quality versus code coverage.

**Comment by nmote** [comment]
Rules already have a `category` metadata field, for example: https://github.com/returntocorp/semgrep-rules/blob/develop/typescript/react/best-practice/react-find-dom.yaml#L14

Is your suggestion that the CLI provides some functionality by which findings can be grouped by category, or you can ask Semgrep to only run rules in a particular category?

**Comment by n614cd** [comment]
I am not using the CLI, so I have not considered that aspect. 
In terms of category, that might work. The problem is the overlap in terms, many of the terms I see in default list (Caching, Correctness, Portability, Performance...) overlap with code quality checks. e.g. in Java string + string + string; is much slower than new StringBuffer(string).append(string).append(string);

It is the overlapping and mix/match of many of the same terminology which is why I would prefer a separate type of finding; especially from a project management perspective. How you manage code quality versus security is completely different.

**Comment by nmote** [comment]
> I am not using the CLI

Okay, so this is feedback about how things are displayed in the app

**Comment by aryx** [comment]
@raghavjain3 do you want to chime in?

**Comment by raghavjain3** [comment]
@n614cd thanks for the feedback! Completely agreed with you on Code Quality being very different than Security findings.

We don't have short-term plans to make the separation in the App more clear (aside from the Category filter we have today) but I'd better understand how you're using Semgrep and what improvements we could make to make the product better.

If you're in our Community Slack, mind sending me a DM and we can go from there?

**Comment by n614cd** [comment]
DM Sent
